### PR TITLE
Fix wrong `VSYNC` event

### PR DESCRIPTION
**Pull request by fzyzcjy**
Firstly, we know `VSYNC` event in timeline is special - `chrome://tracing` tool will show "zebra" colors (gray and white) at every VSYNC. Therefore, it is critical to let this event have correct timing, otherwise every user is doing reasoning with the wrong vsync time.

In the image below, the "a" shows the new VSYNC with this PR, while the "b" shows the old VSYNC interval before this PR. As we can see, the left side of "a" and "b" does not coincide. In other words, before this PR (where we have "b" as the VSYNC and there is no "a"), we consider the wrong time as the vsync time.

The cause is quite simple: Before this PR, the left edge of "VSYNC" event is (for example) the call time of `VsyncWaiterAndroid::OnVsyncFromJava`. However, there exist "frame delay" (e.g. `frameDelayNanos` argument in `OnVsyncFromJava`), so the real vsync time should minus that delay.

As a side remark, in the image below the difference is not very much, but in real scenarios, I have seen once in a while it has large differences. Then you know, the visualization goes wild, and it took me some time before I realized, it is not a bug in code anywhere, but a bug of the VSYNC event time.

image

Close #113475

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by chinmaygarde** [comment]
cc @dnfield

File: shell/common/vsync_waiter.cc [hunk 1]
```diff
@@ -109,6 +109,17 @@ void VsyncWaiter::FireCallback(fml::TimePoint frame_start_time,
     return;
   }
 
+  // The event named "VSYNC" is special in `chrome://tracing` tool - it will
```
Comment (dnfield) [inline shell/common/vsync_waiter.cc:112]: Why here?

This means the event won't get emitted on vsyncs where there's no callback, right?

I think making this more accurate is a good idea, but probably should fire unconditionally

**Comment by dnfield** [comment]
@iskakaushik can you provide a secondary review here?

File: shell/common/vsync_waiter.cc (hunk 1 above)
Comment (fzyzcjy) [reply shell/common/vsync_waiter.cc:112]: You are right! I have not thought about this problem before. Let me change it (probably within hours)

**Review by dnfield — APPROVED** [review]
LGTM, would appreciate a review from @iskakaushik too

**Review by iskakaushik — APPROVED** [review]
LGTM

**Pull request merged**